### ISSUE TITLE
m3u_to_playlist.py - "http.client.RemoteDisconnected: Remote end closed connection without response"

### ISSUE THREAD

**Issue by mmiller7** [Casvt/Plex-scripts#184]
In which file is the bug?
m3u_to_playlist.py

Describe the setup (arguments and their values, how it was run, etc.):
Run on my headless PMS server (Ubuntu Server 22.04.2) in working-directory of the m3u file using: `python3 /mnt/plexmedia/m3u_to_playlist.py -l Music -f Country.m3u`
M3U file is stored on my headless PMS server host

```
plex_ip = '192.168.1.222'
plex_port = '32400'
plex_api_token = 'uxxxxxxxxxxxxxxxxxxxxxxxs'
```
(API token fetched viewing XML capturing value of URL parameter on my laptop web-plex)

Describe what the bug is:
Error with stack trace

```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 445, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 440, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.10/http/client.py", line 1374, in getresponse
    response.begin()
  File "/usr/lib/python3.10/http/client.py", line 318, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.10/http/client.py", line 287, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
http.client.RemoteDisconnected: Remote end closed connection without response

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 755, in urlopen
    retries = retries.increment(
  File "/usr/lib/python3/dist-packages/urllib3/util/retry.py", line 532, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/lib/python3/dist-packages/six.py", line 718, in reraise
    raise value.with_traceback(tb)
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 445, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 440, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.10/http/client.py", line 1374, in getresponse
    response.begin()
  File "/usr/lib/python3.10/http/client.py", line 318, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.10/http/client.py", line 287, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/mnt/plexmedia/m3u_to_playlist.py", line 56, in <module>
    response = m3u_to_playlist(ssn=ssn, library_name=args.LibraryName, file_path=args.File)
  File "/mnt/plexmedia/m3u_to_playlist.py", line 32, in m3u_to_playlist
    sections = ssn.get(f'{base_url}/library/sections').json()['MediaContainer']['Directory']
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 555, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

```

Describe what _should_ be happening:
Importing my m3u into Plex

**Comment by Casvt** [comment]
Without trying anything out yet, I want to say that the importing of m3u files to Plex is a very vague area where it's not allowed in the web UI anymore but still is using the API. It could be the case that Plex removed support for it and that that is the reason that it won't work.

The error suggests that the server actively decided to terminate the request (instead of it just failing for example).

**Comment by mmiller7** [comment]
Found https://github.com/Casvt/Plex-scripts/issues/146 which suggests changing HTTP to HTTPS which sounds plausible, but then it pukes due to the certificate name mismatch.

I'm not proficient enough at Python at the moment to know if there's some flag/change to force accepting invalid SSL certificates but I wonder if that may help.  I may also try seeing if I can figure out my PMS SSL cert common name that Plex assigns it and try that.

**Comment by mmiller7** [comment]
Update - got rid of the errors, but now it also still I don't think worked.

Had to go back to the XML where I got the token and pick up that domain name (`192-168-1-222.xxxxxxxxxxxxxxxxxxxxxx.plex.direct`) for the server name, and changed line 24 to `https`

Now I get no errors but I don't think its working either (can't find playlist)

```
$ python3 /mnt/plexmedia/m3u_to_playlist.py -l Music -f Country.m3u
$ 
```

I'm not sure if there's some verbose/debug mode I can post to help but for the initial error that helped.

**Comment by Casvt** [comment]
I remember that the url you're using actually doesn't work. Try using `192.168.1.222` instead of `192-168-1-222.xxxxxxxxxxxxxxxxxxxxxx.plex.direct`.

**Comment by mmiller7** [comment]
> I remember that the url you're using actually doesn't work. Try using `192.168.1.222` instead of `192-168-1-222.xxxxxxxxxxxxxxxxxxxxxx.plex.direct`.

So that's how I started, but it won't accept the SSL certificate mismatch then.  Is there a way to make it accept/ignore bad certs and continue anyway?

**Comment by Casvt** [comment]
But if you use `http://192.168.1.222`, there shouldn't be any problem with the SSL sert because you're not using Https, instead using http.

**Comment by mmiller7** [comment]
> But if you use `http://192.168.1.222`, there shouldn't be any problem with the SSL sert because you're not using Https, instead using http.

They're all on SSL since they are shared out over the internet I can't reasonably turn off SSL

**Comment by mmiller7** [comment]
UPDATE: Finally got it!  Apparently Plex doesn't support the relative-path m3u all my other devices use. which I didn't realize was the issue with no output and no results.  Once I did a find+replace to change the relative-path to the full library file paths in all my m3u's I was able to import them.

Ultimately, using HTTPS and the `192-168-1-222.xxxxxxxxxxxxxxxxxxxxxx.plex.direct` address was what worked.

If I may make a recommendation, it would be good to add a variable for "protocol" at the top and allow the user to input `http` or `https` along with the other parameters, and put a note to use the plex-direct address if they use SSL.

Finally have all my playlists imported correctly, thank you!
